app/src/main/java/com/hedvig/app/feature/home/ui/changeaddress/GetAddressChangeStoryIdUseCase.kt

```diff
@@ -1,57 +1,60 @@
 import com.hedvig.app.util.apollo.safeQuery
 
 class GetAddressChangeStoryIdUseCase(
     private val createQuoteCartUseCase: CreateQuoteCartUseCase,
     private val apolloClient: ApolloClient,
 ) {
 
-    suspend operator fun invoke(): SelfChangeEligibilityResult {
-        return (
-            apolloClient.query(ActiveContractBundlesQuery()).safeQuery().toEither().flatMap { data ->
-                val storyId = data
-                    .activeContractBundles
-                    .firstOrNull()
-                    ?.angelStories
-                    ?.addressChangeV2
-                if (storyId != null) {
-                    addQuoteCartId(storyId).mapLeft { QueryResult.Error.GeneralError(it.message) }
-                } else {
-                    Either.Right(null)
-                }
-            }.fold(
-                ifLeft = { SelfChangeEligibilityResult.Error(it.message) },
-                ifRight = { maybeStoryId ->
-                    if (maybeStoryId != null) {
-                        SelfChangeEligibilityResult.Eligible(maybeStoryId)
-                    } else {
-                        SelfChangeEligibilityResult.Blocked
-                    }
-                }
-            )
-            )
+    suspend fun invoke(): SelfChangeEligibilityResult {
+        return either<SelfChangeFailureType, SelfChangeEligibilityResult.Eligible> {
+            val activeContractBundlesQueryData = apolloClient.query(ActiveContractBundlesQuery())
+                .safeQuery()
+                .toEither(SelfChangeFailureType::Error)
+                .bind()
+            val storyId =
+                activeContractBundlesQueryData.activeContractBundles.firstOrNull()?.angelStories?.addressChangeV2
+            ensureNotNull(storyId) { SelfChangeFailureType.Blocked }
+            val storyIdWithQuoteCartId = addQuoteCartId(storyId).bind()
+            SelfChangeEligibilityResult.Eligible(storyIdWithQuoteCartId)
+        }.fold(SelfChangeFailureType::toSelfChangeEligibilityResult, ::identity)
     }
 
-    private suspend fun addQuoteCartId(storyId: String): Either<ErrorMessage, String> {
-        return createQuoteCartUseCase.invoke().map { quoteCartId ->
-            appendQuoteCartId(storyId, quoteCartId.id)
-        }
+    private suspend fun addQuoteCartId(storyId: String): Either<SelfChangeFailureType.Error, String> {
+        return createQuoteCartUseCase.invoke()
+            .bimap(
+                { errorMessage -> SelfChangeFailureType.Error(errorMessage.message) },
+                { quoteCartId -> appendQuoteCartId(storyId, quoteCartId.id) }
+            )
     }
 
     sealed class SelfChangeEligibilityResult {
         data class Eligible(val embarkStoryId: String) : SelfChangeEligibilityResult()
         object Blocked : SelfChangeEligibilityResult()
         data class Error(val message: String?) : SelfChangeEligibilityResult()
     }
+
+    private sealed interface SelfChangeFailureType {
```

Comment:
This was basically needed since we don't have Union types and I have to provide something for the `Either` in line 19. I did this to preserve that the invoke function returns a SelfChangeEligibilityResult instead of an Either to not have to change all the call sites as well.
We could even make `invoke` directly respond with an either of <Blocked | Error, Eligible> to not have to have both SelfChangeFailureType and SelfChangeEligibilityResult exist.